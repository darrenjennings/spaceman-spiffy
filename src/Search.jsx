// @flow

import React, { Component } from 'react';
import { SingleDatePicker } from 'react-dates';
import moment from 'moment';

import ComicImage from './ComicImage';
import data from '../data.json';

class Search extends Component {
  state = {
    searchTerm: '',
    timeout: setTimeout(() => {}),
    images: [],
    date: moment('1985-11-18'),
    minimumDate: moment('1985-11-18'),
    maximumDate: moment('1995-12-31'),
    focused: false,
    useDate: false,
    imageCollectionLabel: ''
  };

  props: {
    debounce: number
  };

  updateImages = (text: string) => {
    this.setState({ images: this.filterImages(text, data.images) });
  };

  filterImages = (text: string, images: Array<Comic>) => {
    const imgsFiltered = images
      .filter(image => (this.state.date == null ? true : image.date === this.state.date.format('YYYYMMDD')))
      .filter(
        image =>
          this.state.searchTerm === '' ? true : `${image['alt-text']}`.toUpperCase().indexOf(text.toUpperCase()) >= 0
      );
    const sliced = imgsFiltered.slice(0, 10);
    this.setState({ imageCollectionLabel: `${sliced.length}/${imgsFiltered.length}` });
    return sliced;
  };

  handleDatePicked = (date: moment$Moment) => {
    this.setState({ date }, this.updateImages(''));
  };

  handleDateToggle = () => {
    this.setState({ useDate: !this.state.useDate, searchTerm: '' }, () => {
      if (this.state.useDate) {
        this.setState({ date: this.state.minimumDate }, this.updateImages(this.state.searchTerm));
      } else {
        this.setState({ date: null }, this.updateImages(this.state.searchTerm));
      }
    });
  };

  handleSearchTermChange = (event: SyntheticKeyboardEvent & { target: HTMLInputElement }) => {
    const text = event.target.value;
    const bounceDelay = this.props.debounce;

    this.setState({ searchTerm: text, date: null }, () => {
      clearTimeout(this.state.timeout);
      this.setState({
        timeout: setTimeout(() => {
          this.setState({ images: this.filterImages(text, data.images) });
        }, bounceDelay)
      });
    });
  };

  render() {
    return (
      <div id="search">
        <div className="ui toggle checkbox padding10">
          <input
            type="checkbox"
            name="use_date"
            id="use_date"
            tabIndex="0"
            className="hidden"
            onChange={this.handleDateToggle}
            value={this.state.useDate ? 'on' : 'off'}
          />
          <label htmlFor="use_date">Filter By Date</label>
        </div>
        <div style={{ display: this.state.useDate ? '' : 'none' }}>
          <SingleDatePicker
            numberOfMonths={1}
            isOutsideRange={day => day.isBefore(this.state.minimumDate) || day.isAfter(this.state.maximumDate)}
            date={this.state.date}
            onDateChange={this.handleDatePicked}
            focused={this.state.focused}
            onFocusChange={({ focused }) => this.setState({ focused })}
          />
        </div>
        <div className="ui fluid massive icon input" style={{ display: !this.state.useDate ? '' : 'none' }}>
          <input
            type="text"
            id="search__input"
            placeholder="Search"
            value={this.state.searchTerm}
            onChange={this.handleSearchTermChange}
          />
          <i className="search icon" />
        </div>
        <div
          className="padding10"
          style={{ display: this.state.useDate || this.state.searchTerm === '' ? 'none' : '' }}
        >
          <em>
            Displaying {this.state.imageCollectionLabel} Results with current search term {`"${this.state.searchTerm}"`}
          </em>
        </div>
        {this.state.images.length > 0
          ? this.state.images.map(image => <ComicImage key={image.date} image={image} />)
          : ''}
      </div>
    );
  }
}

export default Search;
