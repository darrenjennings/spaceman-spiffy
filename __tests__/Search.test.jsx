import React from "react";
import { shallow } from "enzyme";
import Search from "../src/Search";
import preload from "../data.json";

test("Search renders correctly", () => {
  const component = shallow(<Search />);
  expect(component).toMatchSnapshot();
});

test("Search should render correct amount of comics", () => {
  const component = shallow(<Search />);
  expect(component.find("input").length).toEqual(1);
});

test("Search should render correct amount of comics based on search term", () => {
  const searchWord = "dad";
  const component = shallow(<Search />);
  component.find("input").simulate("change", { target: { value: searchWord } });
  const showCount = preload.images
    .filter(
      image =>
        `${image["alt-text"]} ${image.image_txt}`
          .toUpperCase()
          .indexOf(searchWord.toUpperCase()) >= 0
    )
    .slice(0, 4).length;
  expect(component.find("img").length).toEqual(showCount);
});
