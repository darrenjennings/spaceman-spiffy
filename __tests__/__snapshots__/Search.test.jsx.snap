// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders correctly 1`] = `
<div
  id="search"
>
  <div
    className="ui toggle checkbox padding10"
  >
    <input
      className="hidden"
      id="use_date"
      name="use_date"
      onChange={[Function]}
      tabIndex="0"
      type="checkbox"
      value="off"
    />
    <label
      htmlFor="use_date"
    >
      Filter By Date
    </label>
  </div>
  <div
    style={
      Object {
        "display": "none",
      }
    }
  >
    <SingleDatePicker
      anchorDirection="left"
      customCloseIcon={null}
      customInputIcon={null}
      date={"1985-11-18T08:00:00.000Z"}
      daySize={39}
      disabled={false}
      displayFormat={[Function]}
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={false}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      id="date"
      initialVisibleMonth={null}
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPrev={null}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      orientation="horizontal"
      phrases={
        Object {
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month",
          "jumpToPrevMonth": "Move backward to switch to the previous month",
          "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
          select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
          "moveFocusByOneMonth": "Switch months",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
          "openThisPanel": "Open this panel",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field",
          "selectFocusedDate": "Select the date in focus",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      renderCalendarInfo={null}
      renderDay={null}
      renderMonth={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      withFullScreenPortal={false}
      withPortal={false}
    />
  </div>
  <div
    className="ui fluid massive icon input"
    style={
      Object {
        "display": "",
      }
    }
  >
    <input
      id="search__input"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
    <i
      className="search icon"
    />
  </div>
  <div
    className="padding10"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <em>
      Displaying 
       Results with current search term 
      ""
    </em>
  </div>
</div>
`;
